### DISCUSSION TITLE
Why does Rye automatically create src/<package_name> directory?

### DISCUSSION THREAD

**Discussion by HeerlijkeHeer** [astral-sh/rye#799] [Q&A] [answered]
Why does Rye automatically create a `src/<package_name>` directory? 

If there is always only going to be a single package inside of `src`, why not only use `src` (i.e. without a package) or `<package_name>`? This is how I see many projects do it. 
If it is an option to have multiple packages under `src`, how and in which cases would you do this? 

That said, I really like working with Rye's `src/<package_name>` layout. It makes it very easy to package and work with my application. However, I need to fully understand why it is this way. The primary reason for asking this question is to grow my understanding. I assume it's some "design pattern" that I don't know about.

Thank you for any information you can offer!

**Comment by bluss** [comment, marked as answer]
I think it's because of https://rye-up.com/philosophy/ - Rye wants to set you off on a good path with good defaults, and this src directory layout helps you do editable installs (pip feature that rye uses by default) correctly.

> Rye shall always create a preferred folder structure for new projects

**Comment by Aran-Fey** [comment]
Is it safe to remove the `src` directory or does `rye` rely on it being there?

**Reply by rmarquis** [reply]
You can use the `bypass-selection` option of the hatch backend.

**Reply by bluss** [reply]
@Aran-Fey as @rmarquis hints, it depends on the configuration of your package. Rye uses established tools for defining the python package (tools such as the pyproject.toml file and by default rye uses the **hatchling** build backend).

So the long answer is as long as your build backend configuration and python package definition is ok with it, you can remove the src directory. One way to test is if rye build still works, and if pip can install your python project.

From the default configuration - that is what rye init creates - it creates a default configuration for `tool.hatch.build.targets.wheel.packages`. You should correct this configuration if you remove src.

**Reply by Aran-Fey** [reply]
Thanks for the info. I took the risk of removing the `src` directory, and so far everything seems to work fine still. I didn't even have to modify my `pyproject.toml`.

That said, I haven't used `rye` very much yet, so there's still a decent chance that some commands (like `rye version` maybe) could lead to problems.
